lib/core/src/chain/liquid.rs

```diff
@@ -60,52 +61,112 @@ pub trait LiquidChainService: Send + Sync {
 }
 
 pub(crate) struct HybridLiquidChainService {
-    electrum_client: ElectrumClient,
-    tip_client: Mutex<ElectrumClient>,
+    client: OnceLock<Client>,
+    config: Config,
 }
 
 impl HybridLiquidChainService {
     pub(crate) fn new(config: Config) -> Result<Self> {
-        let electrum_url = ElectrumUrl::new(&config.liquid_electrum_url, true, true)?;
-        let electrum_client =
-            ElectrumClient::with_options(&electrum_url, ElectrumOptions { timeout: Some(3) })?;
-        let tip_client =
-            ElectrumClient::with_options(&electrum_url, ElectrumOptions { timeout: Some(3) })?;
         Ok(Self {
-            electrum_client,
-            tip_client: Mutex::new(tip_client),
+            config,
+            client: OnceLock::new(),
         })
     }
+
+    fn get_client(&self) -> Result<&Client> {
+        if let Some(c) = self.client.get() {
+            return Ok(c);
+        }
+        let electrum_url = ElectrumUrl::new(&self.config.liquid_electrum_url, true, true)?;
+        let client = electrum_url.build_client(&ElectrumOptions { timeout: Some(3) })?;
+
+        let client = self.client.get_or_init(|| client);
+        Ok(client)
+    }
 }
 
 #[async_trait]
 impl LiquidChainService for HybridLiquidChainService {
     async fn tip(&self) -> Result<u32> {
-        Ok(self.tip_client.lock().unwrap().tip()?.height)
+        let client = self.get_client()?;
+        let mut maybe_popped_header = None;
+        while let Some(header) = client.block_headers_pop_raw()? {
+            maybe_popped_header = Some(header)
+        }
+
+        let new_tip: Option<u32> = match maybe_popped_header {
+            Some(popped_header) => Some(popped_header.height.try_into()?),
+            None => {
+                // https://github.com/bitcoindevkit/rust-electrum-client/issues/124
+                // It might be that the client has reconnected and subscriptions don't persist
+                // across connections. Calling `client.ping()` won't help here because the
+                // successful retry will prevent us knowing about the reconnect.
+                if let Ok(header) = client.block_headers_subscribe_raw() {
+                    Some(header.height.try_into()?)
+                } else {
+                    None
+                }
+            }
+        };
+
+        let tip: u32 = new_tip.ok_or_else(|| anyhow!("Failed to get tip"))?;
```

Comment:
Nit: Same as above. The name `new_tip` doesn't fit. We can return within the `match`.